services/src/main/java/org/keycloak/protocol/oidc/grants/JWTAuthorizationGrantType.java

```diff
@@ -0,0 +1,150 @@
+
+    private static final Logger logger = Logger.getLogger(JWTAuthorizationGrantType.class);
+
+    @Override
+    public Response process(Context context) {
+        setContext(context);
+
+        String assertion = formParams.getFirst(OAuth2Constants.ASSERTION);
+        String expectedAudience = Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName());
+        JWTAuthorizationGrantValidationContext authorizationGrantContext = new JWTAuthorizationGrantValidationContext(assertion, client, expectedAudience);
+
+        try {
+            //client must be confidential
+            authorizationGrantContext.validateClient();
+
+            //validate assertion claim (grant_type already validated to select the grant type)
+            authorizationGrantContext.validateAssertionParameters();
+
+            //validate token is JWT and is valid (the signature is validated by the idp)
+            authorizationGrantContext.validateJWTFormat();
+            authorizationGrantContext.validateTokenActive();
+
+            //mandatory claims
+            authorizationGrantContext.validateAudience();
+            authorizationGrantContext.validateIssuer();
+            authorizationGrantContext.validateSubject();
+
+            //select the idp using the issuer claim
+            String jwtIssuer = authorizationGrantContext.getIssuer();
+            IdentityProviderModel identityProviderModel = getAuthorizationGrantIdentityProviderModel(context.getSession(), jwtIssuer);
+            if (identityProviderModel == null) {
+                throw new RuntimeException("No Identity Provider for provided issuer");
+            }
+            JWTAuthorizationGrantProvider jwtAuthorizationGrantProvider = (JWTAuthorizationGrantProvider) IdentityBrokerService.getIdentityProvider(session, identityProviderModel.getAlias());
+
+            //validate the JWT assertion and get the brokered identity from the idp
+            BrokeredIdentityContext brokeredIdentityContext = jwtAuthorizationGrantProvider.validateAuthorizationGrantAssertion(authorizationGrantContext);
+            if (brokeredIdentityContext == null) {
+                throw new RuntimeException("Error validating JWT with identity provider");
+            }
+
+            //apply identity provider mappers
+            Set<IdentityProviderMapperModel> mappers = session.identityProviders().getMappersByAliasStream(identityProviderModel.getAlias())
+                    .collect(Collectors.toSet());
+            KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
+            for (IdentityProviderMapperModel mapper : mappers) {
+                IdentityProviderMapper target = (IdentityProviderMapper)sessionFactory.getProviderFactory(IdentityProviderMapper.class, mapper.getIdentityProviderMapper());
+                target.preprocessFederatedIdentity(session, realm, mapper, brokeredIdentityContext);
+            }
```

Comment:
Maybe better to execute the mappers in the `validateAuthorizationGrantAssertion` inside the IdP?